File: agents/scala/fw-agent/src/main/scala/afl/agent/model/ServerDocument.scala
package fw.agent.model

import org.mongodb.scala.bson.{BsonArray, BsonNull, BsonValue, Document}
import org.mongodb.scala.documentToUntypedDocument

import scala.jdk.CollectionConverters.*

/** Handler statistics for a server. */
case class HandledCount(handler: String, handled: Int, notHandled: Int)

/** Represents a server document from the AFL servers collection. */
case class ServerDocument(
    uuid: String,
    serverGroup: String,
    serviceName: String,
    serverName: String,
    serverIps: Seq[String],
    startTime: Long,
    pingTime: Long,
    topics: Seq[String],
    handlers: Seq[String],
    handled: Seq[HandledCount],
    state: String,
    manager: String,
    error: Option[Document]
)

object ServerDocument:
  /** Deserialize a MongoDB document into a ServerDocument. */
  def fromBson(doc: Document): ServerDocument =
    val underlying: org.bson.Document = doc
    ServerDocument(
      uuid = underlying.getString("uuid"),
      serverGroup = underlying.getString("server_group"),
      serviceName = underlying.getString("service_name"),
      serverName = underlying.getString("server_name"),
      serverIps = bsonArrayToStrings(doc, "server_ips"),
      startTime = underlying.getLong("start_time"),
      pingTime = underlying.getLong("ping_time"),
      topics = bsonArrayToStrings(doc, "topics"),
      handlers = bsonArrayToStrings(doc, "handlers"),
      handled = doc
        .get[BsonArray]("handled")
        .map(
          _.getValues.asScala.map { v =>
            val d = v.asDocument()
            HandledCount(
              handler = d.getString("handler").getValue,
              handled = d.getInt32("handled").getValue,
              notHandled = d.getInt32("not_handled").getValue
            )
          }.toSeq
        )
        .getOrElse(Seq.empty),
      state = underlying.getString("state"),
      manager = Option(underlying.getString("manager")).getOrElse(""),
      error = optionalSubDoc(doc, "error")
    )

  private def bsonArrayToStrings(doc: Document, key: String): Seq[String] =
    doc
      .get[BsonArray](key)
      .map(_.getValues.asScala.map(_.asString().getValue).toSeq)
      .getOrElse(Seq.empty)

  private def optionalSubDoc(doc: Document, key: String): Option[Document] =
    doc.get[BsonValue](key) match
      case Some(bv) if bv.isDocument => Some(Document(bv.asDocument()))
      case _                         => None

  extension (server: ServerDocument)
    /** Serialize to a MongoDB document. */
    def toBson: Document =
      Document(
        "uuid" -> server.uuid,
        "server_group" -> server.serverGroup,
        "service_name" -> server.serviceName,
        "server_name" -> server.serverName,
        "server_ips" -> server.serverIps,
        "start_time" -> server.startTime,
        "ping_time" -> server.pingTime,
        "topics" -> server.topics,
        "handlers" -> server.handlers,
        "handled" -> server.handled.map { h =>
          Document(
            "handler" -> h.handler,
            "handled" -> h.handled,
            "not_handled" -> h.notHandled
          )
        },
        "state" -> server.state,
        "manager" -> server.manager,
        "error" -> server.error.map(_.toBsonDocument).getOrElse(BsonNull())
      )

File: agents/java/fw-agent/src/main/java/afl/agent/model/ServerDocument.java
// Copyright 2025 Ralph Lemke
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package afl.agent.model;

import org.bson.Document;

import java.util.List;

/**
 * Represents a server in the servers collection.
 */
public record ServerDocument(
        String uuid,
        String serverGroup,
        String serviceName,
        String serverName,
        List<String> serverIps,
        long startTime,
        long pingTime,
        List<String> topics,
        List<String> handlers,
        List<HandledStat> handled,
        String state
) {

    /**
     * Converts to a MongoDB document.
     */
    public Document toDocument() {
        Document doc = new Document()
                .append("uuid", uuid)
                .append("server_group", serverGroup)
                .append("service_name", serviceName)
                .append("server_name", serverName)
                .append("server_ips", serverIps)
                .append("start_time", startTime)
                .append("ping_time", pingTime)
                .append("topics", topics)
                .append("handlers", handlers)
                .append("state", state);

        if (handled != null) {
            doc.append("handled", handled.stream()
                    .map(HandledStat::toDocument)
                    .toList());
        }

        return doc;
    }

    /**
     * Represents handler statistics.
     */
    public record HandledStat(
            String handler,
            int handled,
            int notHandled
    ) {
        public Document toDocument() {
            return new Document()
                    .append("handler", handler)
                    .append("handled", handled)
                    .append("not_handled", notHandled);
        }
    }
}
